### PR TITLE
Support PackageIcon attribute & icon caching

### PR DESCRIPTION
**Pull request by iodes**
Summary of the changes (in less than 80 chars)

 * Support PackageIcon attribute added in the latest version of NuGet.
 * If a PackageIcon exists, it is automatically cached on the server.
 * This pull request resolves the [NU5048](https://docs.microsoft.com/en-us/nuget/reference/errors-and-warnings/nu5048) warning.

Addresses https://github.com/loic-sharma/BaGet/issues/123

### CODE REVIEW

File: src/BaGet.Core.Server/Controllers/PackageContentController.cs [hunk 1]
```diff
@@ -79,5 +79,21 @@ public async Task<IActionResult> DownloadReadmeAsync(string id, string version,
...
+                return NotFound();
+            }
+
+            var iconStream = await _content.GetPackageIconStreamOrNullAsync(id, nugetVersion, cancellationToken);
+            if (iconStream == null)
+            {
+                return NotFound();
+            }
+
+            return File(iconStream, "image/png");
```
Comment (loic-sharma) [inline src/BaGet.Core.Server/Controllers/PackageContentController.cs]: The content type `image/png` may be incorrect. Let's copy nuget.org's logic: https://github.com/NuGet/NuGet.Services.Metadata/blob/91f97be8882ec83c3dfd7521ebe62865ab3adb5d/src/Catalog/Icons/IconProcessor.cs#L170-L278

File: src/BaGet.Core.Server/Extensions/IRouteBuilderExtensions.cs [hunk 2]
```diff
@@ -116,6 +116,11 @@ public static IRouteBuilder MapPackageContentRoutes(this IRouteBuilder routes)
                 template: "v3/package/{id}/{version}/readme",
                 defaults: new { controller = "PackageContent", action = "DownloadReadmeAsync" });
 
+            routes.MapRoute(
+                name: Routes.PackageDownloadIconRouteName,
+                template: "v3/package/{id}/{version}/icon",
+                defaults: new { controller = "PackageContent", action = "DownloadIconAsync" });
```
Comment (loic-sharma) [inline src/BaGet.Core.Server/Extensions/IRouteBuilderExtensions.cs]: I recently ported BaGet to ASP.NET Core 3. You will need to change this to:

```csharp
endpoints.MapControllerRoute(
    name: Routes.PackageDownloadIconRouteName,
     pattern: "v3/package/{id}/{version}/icon",
     defaults: new { controller = "PackageContent", action = "DownloadIcon" })
```

File: src/BaGet.Core/BaGet.Core.csproj [hunk 3]
```diff
@@ -11,7 +11,7 @@
     <PackageReference Include="Microsoft.Extensions.Configuration.Abstractions" Version="$(MicrosoftExtensionsPackageVersion)" />
     <PackageReference Include="Microsoft.Extensions.DependencyInjection.Abstractions" Version="$(MicrosoftExtensionsPackageVersion)" />
     <PackageReference Include="Microsoft.Extensions.Options.ConfigurationExtensions" Version="$(MicrosoftExtensionsPackageVersion)" />
-    <PackageReference Include="NuGet.Packaging" Version="$(NuGetPackageVersion)" />
+    <PackageReference Include="NuGet.Packaging" Version="5.3.1" />
```
Comment (loic-sharma) [inline src/BaGet.Core/BaGet.Core.csproj]: Please revert this change. You can bump up the NuGet.Packaging version here:

https://github.com/loic-sharma/BaGet/blob/05bfc61e1ff86517d2e4571f12524275c2355a0d/src/Directory.Build.props#L33

File: src/BaGet.Core/Storage/PackageStorageService.cs [hunk 4]
```diff
@@ -108,6 +111,29 @@ public class PackageStorageService : IPackageStorageService
...
+            // Store the package's icon, if one exists.
+            if (iconStream != null)
+            {
+                _logger.LogInformation(
+                    "Storing package {PackageId} {PackageVersion} icon at {Path}...",
+                    lowercasedId,
+                    lowercasedNormalizedVersion,
+                    iconPath);
+
+                result = await _storage.PutAsync(iconPath, iconStream, IconContentType, cancellationToken);
```
Comment (loic-sharma) [inline src/BaGet.Core/Storage/PackageStorageService.cs]: Same comment as [here](https://github.com/loic-sharma/BaGet/pull/423/files#r368838123): let's try to detect the content type if possible

File: src/BaGet.Protocol/BaGet.Protocol.csproj [hunk 5]
```diff
@@ -10,7 +10,7 @@
   <ItemGroup>
     <PackageReference Include="Microsoft.Extensions.Logging.Abstractions" Version="$(MicrosoftExtensionsPackageVersion)" />
     <PackageReference Include="Newtonsoft.Json" Version="11.0.2" />
-    <PackageReference Include="NuGet.Protocol" Version="$(NuGetPackageVersion)" />
+    <PackageReference Include="NuGet.Protocol" Version="5.3.1" />
```
Comment (loic-sharma) [inline src/BaGet.Protocol/BaGet.Protocol.csproj]: Please revert this change. You can bump up the NuGet.Packaging version here:

https://github.com/loic-sharma/BaGet/blob/05bfc61e1ff86517d2e4571f12524275c2355a0d/src/Directory.Build.props#L33

File: tests/BaGet.Tests/BaGet.Tests.csproj [hunk 6]
```diff
@@ -13,7 +13,7 @@
 
   <ItemGroup>
     <PackageReference Include="Newtonsoft.Json" Version="11.0.2" />
-    <PackageReference Include="NuGet.Protocol" Version="5.0.0-rtm.5856" />
+    <PackageReference Include="NuGet.Protocol" Version="5.3.1" />
```
Comment (loic-sharma) [inline tests/BaGet.Tests/BaGet.Tests.csproj]: This now uses a variable `$(NuGetPackageVersion)`. That is defined here:

https://github.com/loic-sharma/BaGet/blob/05bfc61e1ff86517d2e4571f12524275c2355a0d/tests/Directory.Build.props#L18

**Comment by loic-sharma** [comment]
I apologize for taking so long to look at this. I've left a few minor comments, I'll merge this in once you've address them. Thank you, this is amazing work! 😊

File: src/BaGet.Core.Server/Controllers/PackageContentController.cs (hunk 1 above)
Comment (loic-sharma) [reply src/BaGet.Core.Server/Controllers/PackageContentController.cs]: Some notes:

* I would add a `DetermineContentTypeAsync` to BaGet's [`StreamExtensions`](https://github.com/loic-sharma/BaGet/blob/05bfc61e1ff86517d2e4571f12524275c2355a0d/src/BaGet.Core/Extensions/StreamExtensions.cs). That way you can do:

```csharp
var contentType = iconStream.DetermineContentTypeAsync(cancellationToken);
return File(iconStream, contentType);
```

* When copying the [`DetermineContentType`](https://github.com/NuGet/NuGet.Services.Metadata/blob/91f97be8882ec83c3dfd7521ebe62865ab3adb5d/src/Catalog/Icons/IconProcessor.cs#L227-L230) method from NuGet.org, you can act as if `onlyGallerySupported` is `true`:

```csharp
// Checks are ordered by format popularity among external icons for existing packages
if (ArrayStartsWith(imageData, PngHeader))
{
    return "image/png";
}

if (ArrayStartsWith(imageData, JpegHeader))
{
    return "image/jpeg";
}

return "";
```

**Comment by softlion** [comment]
@loic-sharma you can update the merging branch directly if the changes are so minor, to speed up fix merging.

File: src/BaGet.Core.Server/BaGetUrlGenerator.cs [hunk 7]
```diff
@@ -126,6 +126,22 @@ public string GetPackageManifestDownloadUrl(string id, NuGetVersion version)
...
+            var versionString = version.ToNormalizedString().ToLowerInvariant();
+
+            return _linkGenerator.GetUriByRouteValues(
+                _httpContextAccessor.HttpContext,
+                Routes.PackageDownloadIconRouteName,
+                values: new
+                {
+                    Id = id,
+                    Version = versionString,
+                    Id2 = id,
```
Comment (loic-sharma) [inline src/BaGet.Core.Server/BaGetUrlGenerator.cs]: Looks like `Id2` is unused

**Comment by loic-sharma** [comment]
I've updated your changes with the latest master branch here: https://github.com/loic-sharma/BaGet/pull/469. Feel free to review, and thanks for opening this!

File: src/BaGet.Core.Server/Controllers/PackageContentController.cs (hunk 1 above)
Comment (loic-sharma) [reply src/BaGet.Core.Server/Controllers/PackageContentController.cs]: It turns out using an empty string `""` does not work. I will try `image/xyz` instead.
